tests/sdk/test_parameter.py

```
@@ -0,0 +1,136 @@
+        }
+
+        sam_parameter_values = SamParameterValues(parameter_values)
+        sam_parameter_values.add_default_parameter_values(sam_template)
+        self.assertEqual(expected, sam_parameter_values.parameter_values)
+
+    def test_add_default_parameter_values_must_override_user_specified_values(self):
+        parameter_values = {
+            "Param1": "value1"
+        }
+
+        sam_template = {
+            "Parameters": {
+                "Param1": {
+                    "Type": "String",
+                    "Default": "template default"
+                }
+            }
+        }
+
+        expected = {
+            "Param1": "value1"
+        }
+
+        sam_parameter_values = SamParameterValues(parameter_values)
+        sam_parameter_values.add_default_parameter_values(sam_template)
+        self.assertEqual(expected, sam_parameter_values.parameter_values)
+
+    def test_add_default_parameter_values_must_skip_params_without_defaults(self):
+        parameter_values = {
+            "Param1": "value1"
+        }
+
+        sam_template = {
+            "Parameters": {
+                "Param1": {
+                    "Type": "String"
+                },
+                "Param2": {
+                    "Type": "String"
+                }
+            }
+        }
+
+        expected = {
+            "Param1": "value1"
+        }
+
+        sam_parameter_values = SamParameterValues(parameter_values)
+        sam_parameter_values.add_default_parameter_values(sam_template)
+        self.assertEqual(expected, sam_parameter_values.parameter_values)
+
+
+    @parameterized.expand([
+        # Array
+        param(["1", "2"]),
+
+        # String
+        param("something"),
+
+        # Some other non-parameter looking dictionary
+        param({"Param1": {"Foo": "Bar"}}),
+
+        param(None)
+    ])
+    def test_add_default_parameter_values_must_ignore_invalid_template_parameters(self, template_parameters):
```

Again, if they're invalid can we fail early?